includes/Core/Util/Entity_Factory.php

```diff
@@ -60,8 +60,54 @@ public static function from_context() {
 
 		// Otherwise, run frontend-specific `WP_Query` logic.
 		if ( $wp_the_query instanceof WP_Query ) {
-			return self::from_wp_query( $wp_the_query );
+			$entity = self::from_wp_query( $wp_the_query );
+
+			if ( null === $entity ) {
+				return null;
+			}
+
+			if ( ! defined( 'AMP__VERSION' ) ) {
+				return $entity;
+			}
+
+			$url_parts    = wp_parse_url( $_SERVER['REQUEST_URI'] ); // phpcs:ignore
+			$new_url_tail = '';
+
+			// check if the $url has amp query param.
+			if ( strpos( $url_parts['query'], 'amp' ) !== false ) {
+				$new_url_tail = '?amp=1';
+			}
+
+			// check if the $url has `/amp` in path.
+			if ( '/amp' === substr( $url_parts['path'], -strlen( '/amp' ) ) ) {
+				$new_url_tail = '/amp';
+			}
+
+			// check if the $url has `/amp/` in path.
+			if ( '/amp/' === substr( $url_parts['path'], -strlen( '/amp/' ) ) ) {
+				$new_url_tail = '/amp/';
+			}
```

Comment:
Let's strip the trailing slash from the `$url_parts['path']` using the [`untrailingslashit`](https://developer.wordpress.org/reference/functions/untrailingslashit/) function and check only for `/amp` suffix.

Also, no need to use `-strlen( '/amp' )` because it always returns the same result which is `-4`. You can leave a comment to explain what `-4` means.

Reply:
Done. But wouldn't only checking for non-trailing slashed version and returning non trailing slashedd url cause issues if the original url is trailing slashed? Don't we need to append `/amp` vs `/amp/` depending on trailing slash? 

Cheers.

Reply:
Yes, it's a good point, @kuasha420. I think you are right, let's append a trailing slash if it exists in the original URL.

Reply:
Should I just revert to old logic of checking twice for trailing slash/non-trailing slash version or is there a better way too do it? @eugene-manuilov

Reply:
yes, let's revert it to the old logic and leave a comment that explains why we have it separated.